asapdiscovery-genetics/asapdiscovery/genetics/blast.py

```diff
@@ -0,0 +1,264 @@
+            )
+
+        match_hits = self.parse_pdb_blast_results(matches_df, min_score=min_id_match)
+
+        for i, hits in enumerate(match_hits):
+            if len(hits) == 0:
+                print("The record doesn't have a PDB with high confidence")
+                if i == 0:
+                    raise ValueError(
+                        "The reference sequence MUST have an existing PDB entry. Maybe check the sequence?"
+                    )
+            else:
+                # Return the pdb entry with the max alignment and max resolution (if there are multiple matches)
+                best_pdb_record = self.choose_best_pdb_entry(hits, best_scores[i])
+                filename = prody.fetchPDB(best_pdb_record, folder=results_folder)
+                subprocess.run(["gunzip", filename])
+                pdb_file_record[i].pdb_file = filename
+                pdb_file_record[i].pdb_id = best_pdb_record
+                # If I'm only requesting the reference PDB, there's no need to generate more files.
+                if ref_only:
+                    break
+
+        return pdb_file_record
+
+    @staticmethod
+    def parse_pdb_blast_results(blast_df, min_score):
+        """For a pdb database search extract pdb ids with a minimum score.
+        Outputs are lists because iterables are needed for functions down the pipeline.
+        """
+
+        match_hits = []
+        for q in blast_df["query"].unique():  # Loop over queries
+            hits = {}
+            # for e in range(len(blast_ids)): # Loop over BLAST entries
+            for idx, e in blast_df.loc[blast_df["query"] == q].iterrows():
+                if e["score"] >= min_score:
+                    raw_id = e["ID"].split("|")
+                    pdb_index = raw_id.index("pdb")
+                    pdb_id = raw_id[pdb_index + 1]
+
+                    hits[pdb_id] = {}
+                    hits[pdb_id]["percent_identity"] = e["score"]
+                    hits[pdb_id]["sequence"] = e["sequence"]
+
+            match_hits.append(hits)
+
+        return match_hits
+
+    @staticmethod
+    def request_rcsb_pdbid(pdb_id):
+        """A function to request a protein entry from the rcsb API with a pdb ID"""
+        requestURL = f"https://data.rcsb.org/rest/v1/core/entry/{pdb_id}"
+        r = requests.get(requestURL, headers={"Accept": "application/json"})
+        if not r.ok:
+            r.raise_for_status()
+            sys.exit()
+        return r.json()
+
+    @classmethod
+    def check_resolution(self, pdb_id):
+        fjson = self.request_rcsb_pdbid(pdb_id)
+        return fjson["rcsb_entry_info"]["resolution_combined"][0]
+
+    @classmethod
+    def choose_best_pdb_entry(self, hits, best_match_percent):
+        """If a sequence have different PDB files all with the same alignment match, we choose and retrieve the one with the highest resolution"""
+        max_res = 10  # some unrealistically large number?
```

Comment:
Not for a Cryo structure.

Reply:
Oh ok, I'll change to 100. I guess that should be fine?